### ISSUE TITLE
Logging performance

### ISSUE THREAD

**Issue by djones6** [IBM-Swift/Kitura#842]
At present, adding Log statements to a Kitura application (or to Kitura itself) carries a performance penalty: the String message must be assembled (which may involve interpolation), even if it is immediately thrown away if either no logger is configured, or the logging level is not enabled.

### Proposal
We could augment the API to provide an optional to short-circuit the construction of the message string, as follows:

`Log.debug("This is the \(n)th time you called me")`
to (eg):
`Log.Debug?.debug("This is the \(n)th time you called me")`

### Performance:
I created [a testcase](https://github.com/djones6/Swift-TechEmpower/blob/master/kitura/Sources/) with 5 implementations.  Each returns `Hello, World!` from the `/plaintext` route, but other than the first, also contain an (artificially) expensive log message: `Log.debug("This might be expensive \(arc4random_uniform(100))")`

My log level is only set to `.warning`, so these messages are not emitted, but in the current implementation, the String still has to be created.  I've then avoided the creation of the string with two methods:  using the existing `if Log.isLogging()` facility, and a small change to add an optional which allows a slightly more convenient invocation (but does not break the existing API).

Performance results indicate that the use of the optional does short-circuit the creation of the String, and suggest there is little difference between the existing test and the optional:
```
Implementation | Avg Throughput | Max Throughput | Avg CPU | Avg RSS (kb) | Avg Lat (ms) | Max Lat (ms)
---------------|----------------|----------------|---------|--------------|--------------|--------------
   LoggingNone |        53513.3 |       53934.30 |    99.7 |        20134 |          2.4 |       230.17
  LoggingDebug |         2787.4 |        2808.61 |    90.5 |        25636 |         45.6 |       243.19
 ...ingDebugIf |        44649.0 |       45104.79 |    90.7 |        23889 |          3.2 |       210.77
 ...ngDebugOpt |        45378.4 |       45975.37 |    92.0 |        25932 |          3.3 |        213.8
   LoggingWarn |         2805.8 |        2830.30 |    91.4 |        25096 |         45.3 |       245.24
```
The advantage is rather that instead of writing:
```
if Log.isLogging(.debug) {
    Log.debug("message")
}
```
you can write: `Log.Debug?.debug("message")`
My experimental changes are here;  https://github.com/djones6/LoggerAPI/tree/issue_842

It may be possible to simplify this further if we were prepared to break the existing API.

**Comment by shmuelk** [comment]
There is the Log.isLogging() function that you can use to query the log level in a if statement that wraps the Log.whatever function call.

What you are proposing breaks the Logger API in a big way. This needs to be thought about carefully. This is not necessarily a bad idea. Just needs some thinking.

**Comment by shmuelk** [comment]
BTW Are you sure that in your example the compiler short circuits the parameter "computation"?

**Comment by djones6** [comment]
@shmuelk @vadimeisenbergibm I updated my original comment,  I found a way to achieve roughly what I originally proposed without breaking the existing API.  I've also added some perf numbers to prove that it does indeed short-circuit the String computation.

**Comment by djones6** [comment]
We can also apply this pattern throughout the Kitura and Kitura-net code.  This does have a measurable benefit for Hello World.  Granted, this is a worst case, but enables us to add more detailed logging in the future:
```
Implementation | Avg Throughput | Max Throughput | Avg CPU | Avg RSS (kb) | Avg Lat (ms) | Max Lat (ms)
---------------|----------------|----------------|---------|--------------|--------------|--------------
   LoggingNone |        53513.3 |       53934.30 |    99.7 |        20134 |          2.4 |       230.17
 ...ebugOpt(1) |        45378.4 |       45975.37 |    92.0 |        25932 |          3.3 |        213.8
 ...ebugOpt(2) |        53412.8 |       54109.71 |    99.7 |        21985 |          2.3 |       200.37
 ...ebugOpt(3) |        44848.1 |       45027.76 |    91.4 |        24445 |          3.3 |       205.17

```
(1) is the same data as my original experiment (logging at `.warning`).  It represents constructing but not actually emitting the internal Kitura log messages.
(2) is the same test as (1), but rebuilt with the corresponding changes to Kitura (https://github.com/djones6/Kitura/tree/issue_842) and Kitura-net (https://github.com/djones6/Kitura-net/tree/issue_842).  It eliminates the String construction for messages that are not emitted.
(3) turns on `.verbose` level logging.  The slight reduction in performance compared to (1) is the cost of actually emitting the messages (in this case, one connection accept message for every 100 requests).

**Comment by vadimeisenbergibm** [comment]
@djones6 we can make it shorter: `Log.debug?("This is the \(n)th time you called me")`

I tried the following code to see that the string is not interpolated when `debug` is nil:
```swift
struct Log {
    typealias LogFunction = (String) -> ()
    var debug: LogFunction?
}

var log = Log()

func foo() -> String {
    print("foo called")
    return "foo"
}

print("call log.debug? with log.debug nil")
log.debug?("interpolate foo = \(foo()) at line \(#line)")

log.debug = { print($0) }

print("call log.debug? with log.debug not nil")
log.debug?("interpolate foo = \(foo()) at line \(#line)")
```

Output:
```
call log.debug? with log.debug nil
call log.debug? with log.debug not nil
foo called
interpolate foo = foo at line 19
```

**Comment by djones6** [comment]
@vadimeisenbergibm There are a couple of issues with the approach above: as far as I can tell, you cannot have argument labels, or default values.

Yesterday, @davidungar suggested using `@autoclosure` to wrap the String param in a closure that doesn't get invoked unless called, and therefore gives us a chance to check log level first.

However, this afternoon I determined that the String doesn't get interpolated at all as long as we never reference it.  This explains why the internal log statements don't carry a penalty when no logger has been assigned.  I guess the compiler is able to determine this statically and skip it.

Turns out, the problem is that neither LoggerAPI nor HeliumLogger are checking the log level (and bailing out early).  HeliumLogger formats the message regardless, then checks only before printing the message to the console.  We can add a level check to the Logger functions, so that Logger implementations aren't invoked unless the message should be logged.
